### PR TITLE
Adopt scalar type changes from TCGC

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Adopt latest changes from TCGC.

### CODE REVIEW

File: src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts [hunk 1]
```diff
@@ -32,7 +32,10 @@ export type InputType =
 
 export interface InputPrimitiveType extends InputTypeBase {
     Kind: SdkBuiltInKinds;
+    Name: string;
     Encode?: string; // In TCGC this is required, and when there is no encoding, it just has the same value as kind
+    CrossLanguageDefinitionId: string;
+    BaseType?: InputPrimitiveType;
```
Comment (live1206) [inline src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts:38]: why is this added? it's never set from tcgc
And we don't need to handle this in generator, either.

File: src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts [hunk 2]
```diff
@@ -62,8 +68,11 @@ export interface InputOffsetDateTimeType extends InputDateTimeTypeBase {
 
 export interface InputDurationType extends InputTypeBase {
     Kind: "duration";
+    Name: string;
     Encode: DurationKnownEncoding;
     WireType: InputPrimitiveType;
+    BaseType?: InputDurationType;
```
Comment (live1206) [inline src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts:74]: we are not handling the newly added `BaseType` in generator?

File: src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs [hunk 3]
```diff
@@ -75,23 +89,26 @@ public TypeFactory(OutputLibrary library, Type unknownType)
...
-#pragma warning restore CS0618 // Type or member is obsolete
                 _ => new CSharpType(typeof(object)),
-            },
-            InputDateTimeType dateTimeType => new CSharpType(typeof(DateTimeOffset)),
-            InputDurationType durationType => new CSharpType(typeof(TimeSpan)),
-            _ => throw new InvalidOperationException($"Unknown type: {inputType}")
+            };
+        }
+
+        private static readonly IReadOnlyDictionary<string, CSharpType> _knownPrimitiveTypes = new Dictionary<string, CSharpType>
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Generation/Types/TypeFactory.cs:100]: The key here is the `crossLanguageDefinitionId`.
And based on the principal that we should keep minimum logic in our emitter only - just taking everything TCGC gives us, the `InputPrimitiveTypeKind` should have exactly the same values as `SdkBuiltInTypeKinds`. And in my TCGC part, this `SdkBuiltInTypeKinds` is defined like this:
```typescript
export type SdkBuiltInKinds = Exclude<IntrinsicScalarName, SdkBuiltInKindsExcludes> | "any";

type SdkBuiltInKindsExcludes = "utcDateTime" | "offsetDateTime" | "duration";
```

Therefore, there should never be a situation that we should add a new type kind into this inputPrimitiveTypeKind, unless typespec is updated and decide to add a new kind into `IntrinsicScalarName` which means a new std type is added into typespec.

File: src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts (hunk 2 above)
Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts:74]: We do not take it right now because we do not have a case yet.
We have it here, because the input types should be identical to the type system provided by TCGC

Comment (ArcturusZhang) [reply src/TypeSpec.Extension/Emitter.Csharp/src/type/input-type.ts:74]: For more details, please refer the description of this PR: https://github.com/Azure/typespec-azure/pull/1015

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputPrimitiveType.cs [hunk 4]
```diff
@@ -3,37 +3,59 @@
...
-    internal InputPrimitiveType(InputPrimitiveTypeKind kind, string? encode) : this(kind)
+    internal const string AzureLocationId = "Azure.Core.azureLocation";
+    internal const string ArmIdId = "Azure.Core.armResourceIdentifier";
+    internal const string UuidId = "Azure.Core.uuid";
+    internal const string ETagId = "Azure.Core.eTag";
+    internal const string CharId = "Azure.Core.char";
+    internal const string ContentTypeId = "Azure.Core.contentType";
+    internal const string ResourceTypeId = "Azure.Core.resourceType";
+
+    internal const string ObjectId = "Obsolete.object";
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputPrimitiveType.cs:16]: [This comment](https://github.com/Azure/autorest.csharp/pull/4883#discussion_r1666387698) shares the same reply here.
Previously we are using some special `InputPrimitiveTypeKind` to handle some legacy types from swagger input.
In this PR since we need to keep the `InputPrimitiveTypeKind` identical as `SdkBuiltInTypeKinds` from TCGC, we no longer could add more kinds into the type (or in other words, we should not), therefore here we just construct a new scalar type here internally with a special `crossLanguageDefinitionId`, so that the `TypeFactory` could know which type to return.

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputPrimitiveType.cs [hunk 5]
```diff
@@ -3,37 +3,59 @@
...
+    internal const string RequestMethodId = "Azure.Core.requestMethod";
+
+    internal const string IPAddressId = "Temp.ipAddress";
```
[the remark this answers is not in the document]
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputPrimitiveType.cs:19]: See https://github.com/Azure/autorest.csharp/pull/4883#discussion_r1683960316

For ipAddress, maybe we could change this to share with `ipv4Address` or `ipv6Address`.

These are not decided yet.
